Allow configuration of driver backends for sdl1/sdl2 modules

joolswills opened the pull request:
This replaces the dispmanx module functionality and reworks the way backend configuration is done, including adding support for choosing to use X11 for modules using sdl1 or sdl2. A new supplementary module "backends" handles the configuration (replacing the dispmanx module).

On upgrade /opt/retropie/configs/all/dispmanx.cfg is renamed to /opt/retropie/configs/all/backends.cfg

Existing settings in this file will still work, but will be updated when using the backends configuration.

Previously for some modules that use sdl1, a flag "dispmanx" was set to specify that the code worked with the sdl1 dispmanx driver on videocore/fkms on the RPI.

This has been replaced with the flags "sdl1" "sdl2" and "sdl1-videocore"

The backend module will use the flags to determine which driver backends are available for current system.
For modules such as advmame which use sdl1 on videocore and sdl2 on other systems, the sdl1-videocore flag can be used which will only apply to the rpi1-3 using the legacy drivers.

The setDispmanx call to default a module to the dispmanx driver has been replaced with setBackend, but the setDispmanx function is left for 3rd party module compatibility.

Calls to setDispmanx should be replaced with a call such as:

  setBackend "$md_id" "dispmanx"

But support for dispmanx should be checked in the module - eg

  hasPlatform "dispmanx" && setBackend "$md_id" "dispmanx"

Module hooks for configure_dispmanx_on / conigure_dispmanx_off are no longer used (only used by fuse).
Instead a function _backend_set_ID is called, with the backend to use, and a force parameter.

By default no existing backend configuration is changed when calling setBackend, unless an additional parameter of 1 is added. This is so a module won't overwrite a user's existing configuration.

If a module is set to launch under dispmanx, runcommand will handle it as before, but will also
now handle launching sdl1 and sdl2 modules under X. It will also run the matchbox-window-manager which is required by some software.

If switching a module to launch under X it will check for the correct dependencies. If not installed
the user will be asked to confirm installation.

cmitu commented on scriptmodules/helpers.sh:
```diff
@@ -647,19 +647,37 @@ function addUdevInputRules() {
-    iniConfig "=" "\"" "$configdir/all/dispmanx.cfg"
-    iniSet $mod_id "$status"
-    chown $user:$user "$configdir/all/dispmanx.cfg"
+    ! isPlatform "dispmanx" || return
```
Shouldn't this be 
```
isPlatform "dispmanx" || return
```

joolswills replied:
Yep.

joolswills commented on scriptmodules/emulators/quasi88.sh:
```diff
@@ -13,8 +13,9 @@ rp_module_id="quasi88"
+rp_module_repo="fb dispmanx x11"
 rp_module_section="exp"
-rp_module_flags="dispmanx !mali !kms"
+rp_module_flags="!mali !kms"
```
This isn't ready for review yet btw. Still working on it. Hence draft. But yeah. I'm going through everything currently

joolswills commented:
I've changed some of the mentioned points. But have gone through it now and also fixed a bunch of typos / old code / etc. There is still going to be things I've missed I'm sure. But review away @cmitu :-)

cmitu reviewed (commented):
I looked over the changes and they seem ok, I left a few notes here and there.

One thing I'm not so fond of is the 'backends' naming. The module changes the SDL1/2 video driver, why not use something closely related to this - at least for the user facing dialog ?

I'll do some more tests and see if anythings comes up.

joolswills commented:
Thanks. I will sort the issues.

I went through a few names, but ended up going full circle with backends. It's more than just switching SDL driver as it involves launching X for the x11 driver and wm and there is possibilities to extend this also.

I'll consider some other names but I didn't find anything I like better yet. So feel free to suggest but not sure I fancy renaming everything :-)

joolswills commented on scriptmodules/supplementary/backends.sh:
```diff
@@ -0,0 +1,138 @@
+            default="SDL2 videocore driver"
+        elif isPlatform "kms"; then
+            default="SDL2 KMS driver"
+        elif isPlatform "mali"; then
```
Good point. It was a future expansion thing and I was testing the logic (forcing different platforms) but I'll remove to avoid confusion.

joolswills commented on scriptmodules/supplementary/backends.sh:
```diff
@@ -0,0 +1,138 @@
+
+function _update_hook_backends() {
+    local dispmanx_cfg="$configdir/all/dispmanx.cfg"
+    local backend_cfg="$configdir/all/backend.cfg"
```
Ah yeah. I renamed this I think at some point when switching names (it was backend.sh for a while). Thanks.

cmitu commented:
> I'll consider some other names but I didn't find anything I like better yet. So feel free to suggest but not sure I fancy renaming everything :-)

Yes, naming is hard :).

I've played a bit with various emulators and works fine. There are a few emulators/ports that don't have the flags in order to get configured (i.e.` mame`, `quake3`, `ioquake3`, `minivmac`), but that can be probably added afterwards.

joolswills commented:
Yeah - I've not added many additional emulators over the existing ones that had dispmanx flag. I think it's best to deal with others later, but I'm not going to merge this yet as I need to go over it more still. It was going to be in draft for longer but it got reviewed so quickly I thought I might as well take it out of draft. But I want to do more testing myself still.

I appreciate the fixes and help of course :-)

joolswills commented:
I made a couple of changes including adding x11 plus cursor option. Going to merge. It may need some tweaks and more modules changed to add support.
